### PR TITLE
feat(wormholes): Make wormholes first class citizens

### PR DESCRIPTION
**Pull request by quyykk**
**Feature:** This PR implements the feature request detailed and discussed in issue #5457.

## Feature Details
This PR makes wormholes first class citizens!

- Wormholes are defined using a `wormhole` (see below).
- Wormholes no longer abuse some planet properties for themselves

## Usage Example

Here's how to define a wormhole:

```
wormhole <name>
        "display name" <name>
        mappable
        link <from> <to>
```

- `<name>` the name of the wormhole.
- `display name` the display name for this wormhole ("???" by default).
- `mappable` whether the arrow on the map is shown (for example, Deneb's wormhole doesn't have this)
- `link`: Specifies that a wormhole in `<from>` will transport you to `<to>` (you can have any number of this).

To make a planet a wormhole simply use the `"wormhole"` attribute like so:

```
planet <name>
    wormhole <wormhole>
```

# Implementation Details

Created a new `Wormhole` class to do Wormhole things instead of putting it all in `Planet`.

## Testing Done
Here's all what I've done:

- Load old save file to confirm that any wormholes already visited are preserved.
- Gone through the Ember Waste wormholes to confirm that wormholes still work correctly.
- Made several save steps on a clean save to confirm that wormholes are saved correctly.
- Went through the Deneb wormhole to make sure that still works.
- Landed on Ringworlds to make sure that didn't break.

## Performance Impact

Negligible.

### CODE REVIEW

**Comment by Zitchas** [comment]
This sounds very useful. It would be very helpful to not have wormholes and planets conflated anymore.

**Comment by Terin** [comment]
Could this support giving wormholes a display name? Something like The Eye in the Wanderer campaign being marked as "???" when it is _extremely_ obvious that it's The Eye is a lot. It'd be cool if others could be updated as you learned more about their area, like the Ember Threshold or Wormhole Alpha, but The Eye is the big one.

**Comment by quyykk** [comment]
> Could this support giving wormholes a display name?

Done. That property is called `"pretty name"` because well the wormhole technically already has a name and the display name is just for displaying. Just add `"pretty name" "The Eye"` to the eye wormholes if you want them to be named :)

**Comment by Zitchas** [comment]
Can this "pretty name" be set using events? It would be great to have a wormhole appear as the normal "??" initially, but then have an event change/add its name later on as situations change.

On a similar tangeant, althought not relevant to this particular issue, I've often wanted to have *planets* labeled with ?? until the player lands on them or other criteria are met.

**Comment by quyykk** [comment]
Yeah you can change every property.

**Comment by Terin** [comment]
Looks good! Would be very handy for naming Ember Wastes wormholes throughout the story there, yeah.

>and the display name is just for displaying.

What do you mean by this, though? You've got `name`, which is used in data, then `display name` for how it looks in-game. That's the syntax used with governments, as well, unless there's a third place with a name I'm not considering.

Also seems like it's not compiling on my end, but that could just be an issue with my setup. Hopefully we can get workflows approved for you before too long.

File: source/Cast.h
Comment (petervdmeer) [inline source/Cast.h:16]: Casting should be a rare operation. If we need this class then we are casting too much.

File: source/CoreStartData.cpp
Comment (petervdmeer) [inline source/CoreStartData.cpp:50]: TrueName() was defined as the name to be used in the datafiles, why change to Name() here?

**Comment by quyykk** [comment]
> That's the syntax used with governments, as well

Oh I didn't know. I'll rename it then.

> Also seems like it's not compiling on my end

Are you on macOS? I don't know how to add the additional files I created to the project that's why it fails :(.

File: source/DistanceMap.cpp
Comment (petervdmeer) [inline source/DistanceMap.cpp:187]: I feel that Wormholes are not properly separated off in this PR. Maybe the amount of casts can be reduced by adding an HasValidWormhole() to object (similar to HasValidPlanet())?

**Review by petervdmeer — CHANGES_REQUESTED** [review]
I didn't review the full PR, but I do already have some review comments.

File: source/CoreStartData.cpp
Comment (quyykk) [reply source/CoreStartData.cpp:50]: Cause it was unnecessary. A planet's name is the name used in the datafiles. I could change it back, it doesn't really matter since `TrueName` just calls `Name`.

File: source/DistanceMap.cpp
Comment (quyykk) [reply source/DistanceMap.cpp:187]: I guess I could! Then I can also get rid of the `assert`s I believe if I introduce `HasValidWormhole` and `GetWormhole` and then the same for planets. Thanks!

File: source/Cast.h
Comment (quyykk) [reply source/Cast.h:16]: I don't actually call the function that asserts all that often (maybe like 5 overall). But as stated below, I'll remove them :)

**Comment by Terin** [comment]
>Are you on macOS? I don't know how to add the additional files I created to the project that's why it fails :(.

Nope, Windows 10 in that case, although Linux seems to work just fine.

**Comment by petervdmeer** [comment]
> Are you on macOS? I don't know how to add the additional files I created to the project that's why it fails :(.

There is a CI-job that will generate the updated XCode-project for you (once you get approval for the CI jobs). Just look at the error-message and download the updated XCode-project file from there after the CI jobs have run.

**Comment by Azure3141** [comment]
Would it be possible to add a way to specify the color of the wormhole link? e.g. red wormhole links for the ember waste wormholes, and purple ones for the Pug created ones.

**Comment by quyykk** [comment]
Sure I can do that, but I think it's a bit out of scope for this PR. Maybe create a feature request first to see what others think?

**Comment by Azure3141** [comment]
I think it would fit well in this PR, since you're already overhauling the wormhole system. Adding a coloring feature separately could mean it would have to be redone entirely if this gets merged, and I don't think the feature is too out of scope since you've already added in bonus features like hiding bidirectional arrows without exploring a wormhole both ways.

**Comment by quyykk** [comment]
> Adding a coloring feature separately could mean it would have to be redone entirely if this gets merged

That change is like 10 lines code. If this gets merged it's trivial to rewrite it. Really.

> added in bonus features like hiding bidirectional arrows without exploring a wormhole both ways

You're right; it's hypocritical of me. I'll just say this: Recoloring parts of the map isn't possible right now (making things appear differently on the map) AFAIK, which means that what you're proposing is novel. The 2 things I added (renaming and the bidirectional arrow thingy) are not novel, at least not to the same extent as yours IMO. sorry

File: source/PlayerInfo.cpp
Comment (quyykk) [inline source/PlayerInfo.cpp:160]: This statement is duplicated. I'd rather put refactor this into a function, but I don't know where to put such a function.

File: data/tests/tests_wormhole_navigation.txt
Comment (Terin) [inline data/tests/tests_wormhole_navigation.txt:237]:
```suggestion
		"visited wormhole" "Ember Threshold (Cardea)"
```
Should fix the unit test error

File: source/Wormhole.h
Comment (prophile) [inline source/Wormhole.h:1]: Copypasta

File: data/tests/tests_wormhole_navigation.txt
Comment (quyykk) [reply data/tests/tests_wormhole_navigation.txt]: <3

**Review by tehhowch — CHANGES_REQUESTED** [review]
My biggest issue is with the datafile syntax change. It's not acceptable.

There wasn't much discussion in the other issues about different solutions. I feel we could very easily keep the existing syntax and achieve the desired result (detailed control over how the game handles wormholes) by introducing an optional `wormhole` node.

In this optional node, the user would be able to set the display name, the system traversal order, whether or not a wormhole should be displayed on the map, and so on. This node is optional in that the game would process all planets after files have been loaded and generate defaulted wormhole entries for those that weren't provided by the user.

I also don't agree that we should create the base Landable class. I think it's perfectly fine for some "deep space" wormhole to have a defense fleet, enable the player to land "inside" of it, and so on. To me what defines a wormhole is, _you land in one system and depart in another_. There's not even a requirement that you always exit into the same system from the same entry point.

Obviously a bunch of that is just not possible at the moment, and this PR isn't the place to just go implement all of it. I do think it reasonable to address the display name,  traversal order, and map arrow state, though.

File: source/GameData.cpp
Comment (quyykk) [reply source/GameData.cpp:368]: It's not requiring it. It's how the new wormholes are named. Should I instead hardcode the existing wormhole names? Because I really don't see what else I can do.

File: source/Wormhole.h
Comment (quyykk) [reply source/Wormhole.h:54]: I don't understand what you mean. This is already in the code (whether the "planet"/wormhole has a spaceport).

Comment (tehhowch) [reply source/Wormhole.h:54]: Sure, but not explicitly like you're making it here - more an artifact of how the game chooses landing sites.

Context ref: https://github.com/endless-sky/endless-sky/issues/5311#issuecomment-691045117

File: source/MapDetailPanel.cpp
Comment (quyykk) [reply source/MapDetailPanel.cpp:678]: Which two different collections do you mean?

File: source/Wormhole.h
Comment (quyykk) [reply source/Wormhole.h:54]: Ah so should I remove it then? Because if I do then NPCs won't land on wormholes anymore.

File: source/Wormhole.cpp
Comment (tehhowch) [inline source/Wormhole.cpp:147]: What do we need these systems for, that we cannot get from `links`? Or is keeping it for outsiders more useful than having them consume `links`?

File: source/MapPanel.cpp
Comment (tehhowch) [inline source/MapPanel.cpp:997]: With a root-node approach, there will be a `GameData::Wormholes` collection you would iterate through here, rather than the planets collection.

File: source/MapPanel.cpp
Comment (tehhowch) [inline source/MapPanel.cpp:1005]: This doesn't obey the links ordering.

**Review by tehhowch — COMMENTED** [review]
To build on my comments, the load algorithm I'm envisioning is something like this:
1. Parse content as before. When a named object appears in multiple systems, it will still trip the is-a-wormhole logic.
2. Also load data from a new root-level "wormhole" token into a class object owned by a `Wormholes` collection in GameData
  1. It shall be named the same as the associated planet
    - if no valid `Planet` exists with its name, it shall be marked invalid and a diagnostic issued by CheckReferences
  2. It shall support customizations like what you've done here
  3. It shall support being modified by GameEvents, like what you've implicitly have done here by making it a child node of `planet`
  4. It shall be considered invalid if its link routes contain invalid systems.
3. After data has loaded, and after events have applied, the collection of planet objects shall be inspected for wormholes.
  1. Any wormholes that don't exist in the Wormholes collection shall be added to it:
   - The associated planet will receive a pointer to this new object.
   - "linked" will be appropriately derived from the planet's description.
   - `links` routing table shall be derived from the order of the planet's systems and create a closed loop.
  2. Any auto-added wormholes created on initial load shall trigger a diagnostic, so creators can update their content.
  3. It is optional to issue a diagnostic for wormholes auto-added via events, as these would always trigger for existing pilots.
  4. Event-based changes to wormhole planets should trigger associated changes to the modern wormhole representation, e.g. `remove description`.

So this loading this file
```
# file a
system A
  object A* # meh, a planet
    ...

system C
  object A* # look ma, a wormhole!
    ...

system B
  object A* # ooh it's a circular one
```
Would result in a wormhole collection containing a wormhole object under the name A*, and it would trigger a diagnostic since there was no `planet A*` definition.
```
planet A*
```
would remove the diagnostic, and the resulting wormhole would not appear on the map (empty description). It would have the links table
```
link A C
link C B
link B A
```

I'm sure I missed something in the above description, let me know if something is unclear.

*Thanks for working on this!*

File: data/map.txt
Comment (quyykk) [reply data/map.txt:27034]: > as that definitely isn't backwards compatible

No?

> there's no way to inject the required data into already-processed or pending GameEvents

There's no need to do that. The Eye and the Pug Wormhole work, because the wormhole needs to be added as an object to the systems in question so that you can use them. Without the event, then the planets are just unused.

File: source/Wormhole.cpp
Comment (quyykk) [reply source/Wormhole.cpp:162]: You're right! I'll change the implementation to follow invalid links instead of removing them. Which means that in your example the round-trip will still work. If someone removes A* from B, then going into the wormhole in A will go to B which is invalid so it will go to C (cause B -> C).

File: source/Wormhole.cpp
Comment (quyykk) [reply source/Wormhole.cpp:147]: Good question. This method is indeed unnecessary.

Comment (quyykk) [reply source/Wormhole.cpp:147]: Turns out I do need it lol haha

**Comment by quyykk** [comment]
I changed it like you said (root level wormhole and co). I still need to test event changes, but they should work (I think).

I also made it so that if you don't specify any links, it generates them automatically. What do you think?

**Comment by quyykk** [comment]
I fixed every bug (I hope). I verified that wormholes are correctly autogenerated (including in events), both one-way and normal wormholes.

The only thing missing at this point is some refactoring I want to do and I believe when events are happen for the first time the wormholes don't generate (they do after you reload your save).

**Comment by quyykk** [comment]
This PR is now Ready To Review<sup>TM</sup>. I found several glaring architectural mistakes that tehhowch would have been against and fixed them. The implementation is now pretty straightforward and clean. Nothing to do anymore on my end.

File: data/map.txt
Comment (petervdmeer) [inline data/map.txt:27098]: The `linked` keyword doesn't sound right to me; every wormhole is linked to somewhere (otherwise it wouldn't be a wormhole). Maybe `mappable` (as in "explored and able to be shown on a map") would be a better keyword, since it controls if the link is shown on a map or not.

File: data/wanderer/wanderers.txt
Comment (petervdmeer) [inline data/wanderer/wanderers.txt:879]: Maybe add a newline here? (Probably not needed for the parser, but it just looks a little bit nicer.)

File: source/Planet.cpp
Comment (petervdmeer) [inline source/Planet.cpp:249]: Why the `IsAutogenerated()` tracking and the separate if-statement? The only situation in which `IsAutoGenerated()` would be true is if LoadFromPlanet() was used to create the wormhole, but the only place where that happens is on line 242 just above here. So we can handle all the effects already above (we could even do the check if the description is empty within the LoadFromPlanet function and handle the GenerateLinks also within the LoadFromPlanet).

File: source/Wormhole.cpp
Comment (petervdmeer) [inline source/Wormhole.cpp:26]: Not a planet's description.

File: source/Wormhole.cpp
Comment (petervdmeer) [inline source/Wormhole.cpp:83]: else child.PrintTrace("Something about display name key given without value")

File: source/Planet.cpp
Comment (petervdmeer) [inline source/Planet.cpp:251]: SetLinked() and GenerateLinks() are actually already handled in LoadFromPlanet().

File: source/Wormhole.cpp
Comment (petervdmeer) [inline source/Wormhole.cpp:194]: Would it make sense to make the code in this function part of `LoadFromPlanet()`? (Or at least rename it to `"GenerateLinksFromPlanets()`).

**Review by petervdmeer — CHANGES_REQUESTED** [review]
Sorry for my late review-feedback. The overall change looks good and I'm happy with the change being implemented here.

File: source/Planet.cpp
Comment (quyykk) [reply source/Planet.cpp:249]: `FinishedLoading` can be called multiple times. If a old-school wormhole (aka a planet) is updated, then we also need to update its wormhole object.

File: source/Planet.cpp
Comment (quyykk) [reply source/Planet.cpp:251]: Same comment as above. This is needed when a planet is updated.

File: source/MapPanel.cpp
Comment (tehhowch) [inline source/MapPanel.cpp:1025]: Do we need to check wormhole validity here too?


(+reference semantics)

File: source/Planet.cpp
Comment (tehhowch) [inline source/Planet.cpp:249]: Should this be `else if`, or done before the one-time fallback?

**Review by tehhowch — COMMENTED** [review]
>  I found several glaring architectural mistakes that tehhowch would have complained about and fixed them. 

Indeed there were 😅

File: source/GameData.cpp
Comment (quyykk) [reply source/GameData.cpp]: Currently the syntax to define wormhole is tightly coupled to exactly one planet (`wormhole <planet name>`). That'll be weird if we ever support more than one planet. So, I propose that we decouple the wormhole name from the planet name and add a new attribute to a planet called `wormhole <wormhole>` that binds the given wormhole to the planet.

What do you think?

File: source/Planet.cpp
Comment (quyykk) [reply source/Planet.cpp:242]: That part will only get executed once, since after the first execution `this->wormhole != nullptr`. Calling LoadFromPlanet multiple times is safe yes.

File: source/Wormhole.cpp
Comment (quyykk) [reply source/Wormhole.cpp:101]: Only in Wormhole::IsValid should be fine I think.

File: source/GameData.cpp
Comment (quyykk) [reply source/GameData.cpp]: Ok I have now implemented what I wrote above. Seems like a better solution tbh and it makes it almost trivial to support the multiple planets one wormhole feature.

File: data/map.txt
Comment (tehhowch) [inline data/map.txt:30166]: We should consider demonstrating that the planet and wormhole can now be named separately. This one is a good candidate IMO.

File: source/Planet.cpp
Comment (quyykk) [reply source/Planet.cpp:252]: > Let's also print this deprecation for the event case, too.

I mean that's already the case since if a wormhole is created through an event then it goes through the code above already or do you mean something else?

File: data/map.txt
Comment (quyykk) [reply data/map.txt:30166]: I honestly have no idea what I should name it. Do  you have any recommendations?

File: source/Planet.cpp
Comment (tehhowch) [inline source/Planet.cpp:258]: My prior comment about logging for the event case was to have logs emitted when we hit this else if branch. I'm ok with not, but knowing "when" the branch is hit could be beneficial.

**Review by tehhowch — COMMENTED** [review]
I swear I'll actually boot this soon 😅

File: source/Planet.cpp
Comment (quyykk) [reply source/Planet.cpp:258]: I can provide a warning if you'd like but I'm not sure it would be useful because you already get a warning for auto-generated wormholes and if you fix that error then you won't see the update warnings anyways.

**Comment by Hecter94** [comment]
@tehhowch 

@Azure3141 was looking into adding custom colours to wormhole links. Would it be appropriate to wait until this PR is merged before doing so, or it can it be done separately? 
If you'd prefer he wait, do we have an idea for when this PR might be ready to merge?

File: source/Wormhole.cpp
Comment (tibetiroka) [inline source/Wormhole.cpp:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```
The standard copyright header was changed recently.

File: source/Wormhole.h
Comment (tibetiroka) [inline source/Wormhole.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

**Review by tibetiroka — CHANGES_REQUESTED** [review]

**Review by petervdmeer — APPROVED** [review]

File: source/Wormhole.cpp
Comment (tibetiroka) [inline source/Wormhole.cpp:2]:
```suggestion
Copyright (c) 2021 by quyykk
```

File: source/Wormhole.h
Comment (tibetiroka) [inline source/Wormhole.h:2]:
```suggestion
Copyright (c) 2021 by quyykk
```

**Review by tibetiroka — CHANGES_REQUESTED** [review]

**Pull request merged**
